corehq/apps/userreports/tasks.py

```diff
@@ -277,43 +280,84 @@ def save_document(doc_ids):
-            if failed_config_ids and working_config_ids:
-                AsyncIndicator.objects.filter(pk=indicator.pk).update(
-                    indicator_config_ids=failed_config_ids
-                )
+        with timer:
+            for doc in doc_store.iter_documents(doc_ids):
+                indicator = indicator_by_doc_id[doc['_id']]
+                successfully_processed = _save_document_helper(indicator, doc)
+                if successfully_processed:
+                    processed_indicators.append(indicator.pk)
+                else:
+                    failed_indicators.append(indicator.pk)
 
         AsyncIndicator.objects.filter(pk__in=processed_indicators).delete()
         AsyncIndicator.objects.filter(pk__in=failed_indicators).update(
             date_queued=None, unsuccessful_attempts=F('unsuccessful_attempts') + 1
         )
+
+    datadog_histogram(
+        'commcare.async_indicator.processing_time', timer.duration,
+        tags=[
+            u'config_ids:{}'.format(indicator_config_ids)
+        ]
+    )
+
+
+def _save_document_helper(indicator, doc):
+    eval_context = EvaluationContext(doc)
+    working_config_ids = []
+    failed_config_ids = []
+    for config_id in indicator.indicator_config_ids:
+        adapter = None
+        try:
+            config = _get_config(config_id)
+            adapter = get_indicator_adapter(config, can_handle_laboratory=True)
+            adapter.save(doc, eval_context)
+            eval_context.reset_iteration()
+            working_config_ids.append(config_id)
+        except (DatabaseError, ESError, InternalError, RequestError,
+                ConnectionTimeout, ProtocolError, ReadTimeout):
+            # a database had an issue so don't log it and go on to the next config
+            failed_config_ids.append(config_id)
+        except Exception as e:
+            # getting the config could fail before the adapter is set
+            if adapter:
+                adapter.handle_exception(doc, e)
+            failed_config_ids.append(config_id)
+
+    # only want to change if there were some that worked and failed
+    if failed_config_ids and working_config_ids:
+        AsyncIndicator.objects.filter(pk=indicator.pk).update(
+            indicator_config_ids=failed_config_ids
+        )
+
+    return len(failed_config_ids) == 0
+
+
+@periodic_task(
+    run_every=crontab(minute="*/5"),
+    queue=settings.CELERY_PERIODIC_QUEUE,
+)
+def async_indicators_metrics():
+    for ind in _indicators_by_count():
+        config_ids = ind['indicator_config_ids']
+        count = ind['indicator_config_ids__count']
+        datadog_gauge('commcare.async_indicator.indicator_count', count, tags=config_ids)
+
+
+def _indicators_by_count():
+    """
+    Number of docs in the queue that have a specific set of indicator config ids
+
+    query returns
+    [
+      {"indicator_config_ids": [data_source_id_1, data_source_id_1], indicator_config_ids__count: "12"}
```

Comment:
Note for me tomorrow. Make this return {datasource_id: count} as indicator_config_ids can be out of order